File: Army(c++)/Basic/MagicPower.cpp
#include "MagicPower.hpp"

MagicPower::MagicPower(int mana) : mana(mana), manaLimit(mana) {}

MagicPower::~MagicPower() {}

void MagicPower::ensureHaveIsMana() {
    if ( this->mana == 0 ) {
       throw MagicManaOff();
   }
}

int MagicPower::getMana() const {
    return this->mana;
}

int MagicPower::getManaLimit() const {
    return this->manaLimit;
}

void MagicPower::addMana(int mana) {
    this->ensureHaveIsMana();
    
    if ( mana <= (this->manaLimit - this->mana) ) {
        this->mana += mana;
    } else {
        this->mana = this->manaLimit;
    }
}

void MagicPower::takeMana(int cost) {
    this->ensureHaveIsMana();
    
    this->mana -= cost;
    
    if ( this->mana < 0 ) {
        this->mana = 0;
    }
}

std::ostream& operator<<(std::ostream& out, const MagicPower& magic) {
    out << "Mana " << magic.getMana() << std::endl;
    out << "Mana Limit " << magic.getManaLimit() << std::endl;
    
    return out;
}

File: Army(java)/Basic/MagicPower.java
package Basic;

public class MagicPower {
    protected int mana;
    protected int manaLimit;
    
    public MagicPower(int mana) {
        this.mana = mana;
        this.manaLimit = mana;
    }
        
    public void ensureHaveIsMana() throws MagicManaOffException {
        if ( mana == 0 ) {
            throw new MagicManaOffException();
        }
    }

    public int getMana() {
        return mana;
    }

    public int getManaLimit() {
        return manaLimit;
    }
        
    public void addMana(int newMana) throws MagicManaOffException {
        ensureHaveIsMana();
        
        if ( newMana <= (manaLimit - mana) ) {
            mana += newMana;
        } else {
            mana = manaLimit;
        }
    } 

    public void takeMana(int cost) throws MagicManaOffException {
        ensureHaveIsMana();
        
        mana -= cost;
        
        if ( mana < 0 ) {
            mana = 0;
        }
    }

    public void printPower() {
        System.out.println("Mana\t\t" + " " + mana);
        System.out.println("Mana limit\t" + " " + manaLimit);
    }
}
